### PR TITLE
Conditional access support with Microsoft Graph

### PR DESCRIPTION
**Pull request by liammoat**
This PR builds on #373, and uses https://github.com/manicminer/hamilton, to add support for conditional access policies.

### CODE REVIEW

**Comment by kaovd** [comment]
Now support for graph is merged can we get these linting issues fixed so we're all good to merge? @liammoat

**Comment by alexwilcox9** [comment]
I might be wrong but I don't believe any new features are getting merged until v2.0 or later of the provider (once AAD graph is removed)

**Comment by manicminer** [comment]
That is the case, there are a lot of changes for v2 which need to be merged ahead of any new features, but this PR can stay open in the meantime and we'll take a look closer to the time.

File: internal/services/conditionalAccessPolicies/conditional_access_policy_resource_test.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("id").Exists(),
+				check.That(data.ResourceName).Key("display_name").HasValue(fmt.Sprintf("acctest-CONPOLICY-%d", data.RandomInteger)),
+				check.That(data.ResourceName).Key("state").HasValue("disabled"),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
```
Comment (kaovd) [inline internal/services/conditionalAccessPolicies/conditional_access_policy_resource_test.go:55]:
```suggestion

func TestAccConditionalAccessPolicy_complete(t *testing.T) {
	data := acceptance.BuildTestData(t, "azuread_conditional_access_policy", "test")
	r := ConditionalAccessPolicyResource{}

	data.ResourceTest(t, r, []resource.TestStep{
		{
			Config: r.complete(data),
			Check: resource.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
				check.That(data.ResourceName).Key("id").Exists(),
				check.That(data.ResourceName).Key("display_name").HasValue(fmt.Sprintf("acctest-CONPOLICY-%d", data.RandomInteger)),
				check.That(data.ResourceName).Key("state").HasValue("enabledForReportingButNotEnforced"),
			),
		},
		data.ImportStep(),
	})
}

```

File: internal/services/conditionalAccessPolicies/conditional_access_policy_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  display_name = "acctest-CONPOLICY-%[1]d"
+  state = "enabled"
+
+  conditions {
+    applications {
+      included_applications = ["All"]
+    }
+    users {
+      included_users = ["All"]
+      excluded_users = ["GuestsOrExternalUsers"]
```
Comment (kaovd) [inline internal/services/conditionalAccessPolicies/conditional_access_policy_resource_test.go:89]:
```suggestion
      included_users = ["GuestsOrExternalUsers"]
```

File: internal/services/conditionalAccessPolicies/conditional_access_policy_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    client_app_types = ["mobileAppsAndDesktopClients", "browser"]
+    locations {
+      included_locations = ["All"]
+      excluded_locations = ["AllTrusted"]
+    }
+  }
+
+  grant_controls {
+    operator          = "OR"
+    built_in_controls = ["block"]
```
Comment (kaovd) [inline internal/services/conditionalAccessPolicies/conditional_access_policy_resource_test.go:157]:
```suggestion
  state = "enabledForReportingButNotEnforced"

  conditions {
    applications {
       included_applications = ["00000007-0000-0000-c000-000000000000"]
       excluded_applications = ["00000009-0000-0000-c000-000000000000"]
    }
    users {
      included_users = ["All"]
      excluded_users = ["GuestsOrExternalUsers"]
    }
    client_app_types = ["mobileAppsAndDesktopClients", "browser"]
    locations {
      included_locations = ["All"]
      excluded_locations = ["AllTrusted"]
    }
  }

  grant_controls {
    operator          = "OR"
    built_in_controls = ["mfa"]
```

**Review by kaovd — COMMENTED** [review]
Tested this out a bit and looks good, had to tweak the tests a bit though noticed you added in another test scenario but didnt hook it via a func, changed the test a bit as well as I dont think the excluded Id (ms Graph) exists in cond access so went with Common Data Service as its the first one on the list 
Also saw cannot have both All and application Ids in the included applications list as it will just try read All as an app Id and is just auto removed by azure after you go manually take a look at the policy. 
Additionally changed the first basic test to instead block guests as a block for all users even in a test I'm just a bit wary of incase of lockout.

**Comment by manicminer** [comment]
Hi @liammoat, many thanks for working on this PR and opening it in advance of the 2.0 release! As it looks like #466 is an extension of this work I'm going to go ahead and close this in favor of that one. I'll see what I can do to add your details to the initial commits there for attribution. Thanks again, the contribution is greatly appreciated!

**Pull request closed without merging**